### PR TITLE
Make repack process be able to generate new data tiers: L1SCOUT, HLTSCOUT

### PR DESCRIPTION
**Pull request by haozturk**
#### PR description:

This PR updates the REPACK process such that it is able to generate two new data tiers: `HLTSCOUT` and `L1SCOUT` in addition to `RAW`. I've updated the event content of `HLTSCOUT` and `L1SCOUT` such that they inherit their output commands from `HLTriggerRAW` and `L1TriggerRAW` respectively. Additionally, I've included the output commands of L1SCOUT and HLTSCOUT in the output module of the Repack process.

I've coordinated with @drkovalskyi @germanfgv @dynamic-entropy on this work.

Related tickets: 

- https://its.cern.ch/jira/browse/CMSTZ-1069
- https://its.cern.ch/jira/browse/CMSTRANSF-829

#### PR validation:

Note that I've performed these changes by using "CMSSW_14_0_1" as a base and I've tested them on the same release. The original branch is in [this PR](https://github.com/cms-sw/cmssw/pull/44380). I wanted to rebase my changes into upstream master and re-test, however there has been conflicts during this rebase in files that I haven't changed and don't know how to resolve. The tests I performed and showed below were done in the version that the backport PR uses as source. If anybody can show me how I can test this version, I'm happy to do so. You can find the details of the testing I've performed in the backport PR below:

I used `Configuration/DataProcessing/test/RunRepack.py` for testing. I added a new option to this script

```

$ python3 DataProcessing/test/RunRepack.py --lfn /store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat --data-tier RAW
Now do:
cmsRun -e RunRepackCfg.py
[haozturk@lxplus802 Configuration]$ cmsRun -e RunRepackCfg.py
08-Mar-2024 15:55:15 CET  Initiating request to open file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
08-Mar-2024 15:55:17 CET  Successfully opened file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
Begin processing the 1st record. Run 377350, Event 172727, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:20.650 CET
Begin processing the 2nd record. Run 377350, Event 255073, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:20.739 CET
Begin processing the 3rd record. Run 377350, Event 495959, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:20.825 CET
Begin processing the 4th record. Run 377350, Event 5487, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:20.907 CET
Begin processing the 5th record. Run 377350, Event 5498, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:20.990 CET
Begin processing the 6th record. Run 377350, Event 4244, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:21.075 CET
Begin processing the 7th record. Run 377350, Event 5482, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:21.164 CET
Begin processing the 8th record. Run 377350, Event 4232, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:21.250 CET
Begin processing the 9th record. Run 377350, Event 4252, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:21.329 CET
Begin processing the 10th record. Run 377350, Event 5499, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:21.419 CET
08-Mar-2024 15:55:21 CET  Closed file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
[haozturk@lxplus802 Configuration]$ python3 DataProcessing/test/RunRepack.py --lfn /store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat --data-tier L1SCOUT
Now do:
cmsRun -e RunRepackCfg.py
[haozturk@lxplus802 Configuration]$ cmsRun -e RunRepackCfg.py
08-Mar-2024 15:55:38 CET  Initiating request to open file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
08-Mar-2024 15:55:40 CET  Successfully opened file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
Begin processing the 1st record. Run 377350, Event 172727, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:41.687 CET
Begin processing the 2nd record. Run 377350, Event 255073, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:41.760 CET
Begin processing the 3rd record. Run 377350, Event 495959, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:41.837 CET
Begin processing the 4th record. Run 377350, Event 5487, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:41.905 CET
Begin processing the 5th record. Run 377350, Event 5498, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:41.973 CET
Begin processing the 6th record. Run 377350, Event 4244, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:42.041 CET
Begin processing the 7th record. Run 377350, Event 5482, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:42.108 CET
Begin processing the 8th record. Run 377350, Event 4232, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:42.174 CET
Begin processing the 9th record. Run 377350, Event 4252, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:42.238 CET
Begin processing the 10th record. Run 377350, Event 5499, LumiSection 1 on stream 0 at 08-Mar-2024 15:55:42.305 CET
08-Mar-2024 15:55:42 CET  Closed file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
[haozturk@lxplus802 Configuration]$ 
[haozturk@lxplus802 Configuration]$ 
[haozturk@lxplus802 Configuration]$ python3 DataProcessing/test/RunRepack.py --lfn /store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat --data-tier HLTSCOUT
Now do:
cmsRun -e RunRepackCfg.py
[haozturk@lxplus802 Configuration]$ cmsRun -e RunRepackCfg.py
08-Mar-2024 15:56:00 CET  Initiating request to open file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
08-Mar-2024 15:56:02 CET  Successfully opened file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
Begin processing the 1st record. Run 377350, Event 172727, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.414 CET
Begin processing the 2nd record. Run 377350, Event 255073, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.495 CET
Begin processing the 3rd record. Run 377350, Event 495959, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.577 CET
Begin processing the 4th record. Run 377350, Event 5487, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.656 CET
Begin processing the 5th record. Run 377350, Event 5498, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.737 CET
Begin processing the 6th record. Run 377350, Event 4244, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.818 CET
Begin processing the 7th record. Run 377350, Event 5482, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.903 CET
Begin processing the 8th record. Run 377350, Event 4232, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:03.984 CET
Begin processing the 9th record. Run 377350, Event 4252, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:04.071 CET
Begin processing the 10th record. Run 377350, Event 5499, LumiSection 1 on stream 0 at 08-Mar-2024 15:56:04.160 CET
08-Mar-2024 15:56:04 CET  Closed file root://eoscms.cern.ch//eos/cms/store/t0streamer/Data/PhysicsCommissioning/000/377/350/run377350_ls0001_streamPhysicsCommissioning_StorageManager.dat
[haozturk@lxplus802 Configuration]$ 
[haozturk@lxplus802 Configuration]$ ls -l
total 3195
drwxr-xr-x. 2 haozturk zp   4096 Mar  8 15:51 DataProcessing
drwxr-xr-x. 2 haozturk zp   4096 Mar  8 15:51 EventContent
-rw-r--r--. 1 haozturk zp  14137 Mar  8 15:56 FrameworkJobReport.xml
-rw-r--r--. 1 haozturk zp   8411 Mar  8 15:55 RunRepackCfg.py
-rw-r--r--. 1 haozturk zp 561807 Mar  8 15:56 write_PrimDS1_HLTSCOUT.root
-rw-r--r--. 1 haozturk zp 495347 Mar  8 15:55 write_PrimDS1_L1SCOUT.root
-rw-r--r--. 1 haozturk zp 562065 Mar  8 15:55 write_PrimDS1_RAW.root
-rw-r--r--. 1 haozturk zp 561807 Mar  8 15:56 write_PrimDS2_HLTSCOUT.root
-rw-r--r--. 1 haozturk zp 495347 Mar  8 15:55 write_PrimDS2_L1SCOUT.root
-rw-r--r--. 1 haozturk zp 562065 Mar  8 15:55 write_PrimDS2_RAW.root
[haozturk@lxplus802 Configuration]$ edmDumpEventContent write_PrimDS1_RAW.root 
Type                      Module                   Label     Process   
-----------------------------------------------------------------------
GlobalObjectMapRecord     "hltGtStage2ObjectMap"   ""        "HLT"     
edm::TriggerResults       "TriggerResults"         ""        "HLT"     
trigger::TriggerEvent     "hltTriggerSummaryAOD"   ""        "HLT"     
FEDRawDataCollection      "rawDataCollector"       ""        "LHC"     
[haozturk@lxplus802 Configuration]$ edmDumpEventContent write_PrimDS1_L1SCOUT.root 
Type                     Module               Label     Process   
------------------------------------------------------------------
FEDRawDataCollection     "rawDataCollector"   ""        "LHC"     
[haozturk@lxplus802 Configuration]$ edmDumpEventContent write_PrimDS1_HLTSCOUT.root 
Type                      Module                   Label     Process   
-----------------------------------------------------------------------
GlobalObjectMapRecord     "hltGtStage2ObjectMap"   ""        "HLT"     
edm::TriggerResults       "TriggerResults"         ""        "HLT"     
trigger::TriggerEvent     "hltTriggerSummaryAOD"   ""        "HLT"     
FEDRawDataCollection      "rawDataCollector"       ""        "LHC"  
```

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Backport is https://github.com/cms-sw/cmssw/pull/44380

### CODE REVIEW

**Comment by drkovalskyi** [comment]
Could someone please review it. It's urgently needed for Tier0 to start testing for Run2024 data taking.

File: Configuration/EventContent/python/EventContent_cff.py [hunk 1]
```diff
@@ -184,6 +184,30 @@ def SwapKeepAndDrop(l):
...
+# HLTSCOUT Data Tier definition
+#
+#
+HLTSCOUTEventContent = cms.PSet(
+    outputCommands = cms.untracked.vstring('drop *'),
+    splitLevel = cms.untracked.int32(0),
+    compressionAlgorithm=cms.untracked.string("LZMA"),
+    compressionLevel=cms.untracked.int32(4)
+)
+HLTSCOUTEventContent.outputCommands.extend(HLTriggerRAW.outputCommands)
```
Comment (mmusich) [inline Configuration/EventContent/python/EventContent_cff.py:197]: we've been discussing internally within TSG that this could be just [`HLTScouting`](https://github.com/cms-sw/cmssw/blob/d389c1c21ce701feb341cd00236d5b432160d2b5/HLTrigger/Configuration/python/HLTrigger_EventContent_cff.py#L412) or even better we could add a new `HLTriggerHLTSCOUT` to `HLTrigger_EventContent_cff.py` replacing `HLTScouting` (same `keep` instructions but adding a ` 'drop *_hlt*_*_*'` at the beginning). This would need to be done internally by TSG and can go as a refinement after this PR is integrated.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
urgent
   * > it's urgently needed for Tier0 to start testing for Run2024 data taking.

**Comment by drkovalskyi** [comment]
@mmusich, we were thinking along this line as well, but since it doesn't have a drop statement, we had to proceed with this proposal. Let's start with something that can be used right away and refine later.

**Comment by mmusich** [comment]
@drkovalskyi 

>  Let's start with something that can be used right away and refine later.

agreed.

**Comment by mmusich** [comment]
tagging @cms-sw/l1-l2 in case there are comments on the `L1SCOUT` part.

**Comment by aloeliger** [comment]
I'm going to redirect that question to @dinyar, our scouting expert.

**Comment by aloeliger** [comment]
Otherwise, it looks fine to me.

**Comment by dinyar** [comment]
Hi @aloeliger, thanks for the tag! I'll pass it right on to @Mmiglio who is working on the CMSSW-based parts of the L1T scouting system.

**Comment by Mmiglio** [comment]
Hi, thanks for the tag. It looks good from the L1TScouting perspective as well.

File: Configuration/EventContent/python/EventContent_cff.py [hunk 2]
```diff
@@ -184,6 +184,30 @@ def SwapKeepAndDrop(l):
...
+# L1SCOUT Data Tier definition
+#
+#
+L1SCOUTEventContent = cms.PSet(
+    outputCommands = cms.untracked.vstring('drop *'),
+    splitLevel = cms.untracked.int32(0),
+    compressionAlgorithm=cms.untracked.string("LZMA"),
+    compressionLevel=cms.untracked.int32(4)
+)
+L1SCOUTEventContent.outputCommands.extend(L1TriggerRAW.outputCommands)
```
Comment (Mmiglio) [inline Configuration/EventContent/python/EventContent_cff.py:209]: Shall this be changed in a different PR? The L1Scout event content is different from the L1Trigger

**Comment by davidlange6** [comment]
fundamentally, why does the Tier0 repack need to know about keep or drop statements?

**Comment by drkovalskyi** [comment]
@davidlange6 shouldn't it drop stuff that is not needed for a given data tier?

**Comment by davidlange6** [comment]
Why would the hlt be sending stuff not needed?

On Mar 13, 2024, at 12:41 PM, Dmytro Kovalskyi ***@***.***> wrote:



@davidlange6<https://github.com/davidlange6> shouldn't it drop stuff that is not needed for a given data tier?

**Comment by drkovalskyi** [comment]
HLT in general produces more products than one needs to store. Somewhere we should enforce the content. Maybe it should happen at P5, but if it doesn't, it's just an issue of bandwidth between P5 and Meyrin. While if we let it through Tier0 it will go to tape and space will be wasted long term. Do you see a specific issue?

**Comment by davidlange6** [comment]
It does?  Is there an example somewhere? [certainly not the case for the baseline raw data as you can see from the way the repack step has worked]

> HLT in general produces more products than one needs to store. Somewhere we should enforce the content. Maybe it should happen at P5, but if it doesn't, it's just an issue of bandwidth between P5 and Meyrin. While if we let it through Tier0 it will go to tape and space will be wasted long term. Do you see a specific issue?

**Comment by mmusich** [comment]
> It does?  Is there an example somewhere?

I don't think it does. For the record for the `ScoutingPFRun3` PD (Stream `ScoutingPF`) we have the following

Screenshot from 2024-03-13 12-59-31

what are the surplus collections?

**Comment by drkovalskyi** [comment]
I didn't know that HLT enforces event content on its side, which is a good thing. Still, we are adding new data tiers. So we are adding new event content for them.

**Comment by davidlange6** [comment]
The new data tiers also seem duplicative with the data set names.. but I presume there is a discussion there that I’ve missed.



I didn't know that HLT enforces event content on its side, which is a good thing. Still, we are adding new data tiers. So we are adding new event content for them.

**Comment by drkovalskyi** [comment]
We are introduce new Data Tiers to be able clearly identify scouting data in operations.

**Comment by antoniovilela** [comment]
> urgent
> 
> * > it's urgently needed for Tier0 to start testing for Run2024 data taking.

When it is needed in a production release?

In such cases, please give us a heads up in an ORP meeting.
Thanks.

**Comment by drkovalskyi** [comment]
The sooner the better. We need to have it tested in Tier0 and output validated by the TSG scouting team, but we have some time scouting data will appear in the menu.

**Comment by antoniovilela** [comment]
> The sooner the better. We need to have it tested in Tier0 and output validated by the TSG scouting team, but we have some time scouting data will appear in the menu.

Is next week good for you?

**Comment by drkovalskyi** [comment]
Yes. Thanks.

**Comment by mmusich** [comment]
> What has changed w.r.t. hlt scouting that requires this? I guess that has been repacked for years now, no?

from TSG side nothing has changed. The request comes from offline.

**Comment by antoniovilela** [comment]
@davidlange6 @drkovalskyi @cms-sw/ppd-l2 @cms-sw/orp-l2 

May we move on with this PR, which will add new data tiers for L1/HLT Scouting data?

**Comment by drkovalskyi** [comment]
The change was discussed and approved by L1 of the relevant groups. Details can be fine tuned later.

**Comment by davidlange6** [comment]
as it stands I guess it will write out empty l1scouting data sets and possibly incomplete hltscouting data sets. Maybe thats ok, dunno. But go for it if its not possible to implement the trivial changes discussed above [not that i really understand the need for this pr at all :) ]

File: Configuration/DataProcessing/python/Repack.py [hunk 3]
```diff
@@ -55,12 +67,15 @@ def repackProcess(**args):
...
-            compressionAlgorithm=copy.copy(RAWEventContent.compressionAlgorithm),
-            compressionLevel=copy.copy(RAWEventContent.compressionLevel),
+            compressionAlgorithm=copy.copy(eventContent.compressionAlgorithm),
+            compressionLevel=copy.copy(eventContent.compressionLevel),
             fileName = cms.untracked.string("%s.root" % moduleLabel)
             )
 
-        outputModule.dataset = cms.untracked.PSet(dataTier = cms.untracked.string("RAW"))
+        if dataTier != defaultDataTier:
+            outputModule.outputCommands = copy.copy(eventContent.outputCommands)
```
Comment (davidlange6) [inline Configuration/DataProcessing/python/Repack.py:76]: @drkovalskyi - my suggestion is to remove this line and the previous one, which removes the concern I've raised.

Comment (drkovalskyi) [reply Configuration/DataProcessing/python/Repack.py:76]: I agree. Let's not enforce event content in repacking. It introduces more problems then it solves. We may still need the new EventContent for central production, so I would still keep that part, but what concerns repacking itself, I agree with David's proposal.

Comment (davidlange6) [reply Configuration/DataProcessing/python/Repack.py:76]: they are needed for MC production and/or derived data sets, but those should be correct if defined... ( i guess the one for hlt scouting is already present somewhere as its in miniaod already iiuc.)

Comment (drkovalskyi) [reply Configuration/DataProcessing/python/Repack.py:76]: For MC uses cases we will have time to fine tune and iterate.

Comment (davidlange6) [reply Configuration/DataProcessing/python/Repack.py:76]: please open an issue so that you don't forget to clean things up

Comment (haozturk) [reply Configuration/DataProcessing/python/Repack.py]: I've removed these two lines as you suggested.

**Comment by mmusich** [comment]
please test

File: Configuration/DataProcessing/python/Repack.py (hunk 3 above)
Comment (antoniovilela) [reply Configuration/DataProcessing/python/Repack.py]: > please open an issue so that you don't forget to clean things up

@drkovalskyi

Please let us know once you open the issue.
Thanks.

Comment (drkovalskyi) [reply Configuration/DataProcessing/python/Repack.py]: @haozturk could you please follow up on this request

Comment (haozturk) [reply Configuration/DataProcessing/python/Repack.py]: Here is the issue: https://github.com/cms-sw/cmssw/issues/44409 Please correct me if there is any missing or incorrect statements there.

**Comment by antoniovilela** [comment]
+1
